ops/model.py

```diff
@@ -3489,3 +3525,57 @@ def validate_label_value(cls, label: str, value: str):
         if re.search('[,=]', v) is not None:
             raise ModelError(
                 f'metric label values must not contain "," or "=": {label}={value!r}')
+
+
+class LazyNotice:
+    """Provide lazily-loaded access to a Pebble notice's details.
+
+    The attributes provided by this class are the same as those of
+    :class:`ops.pebble.Notice`, however, the notice details are only fetched
+    from Pebble if necessary (and cached on the instance).
+    """
+
+    id: str
+    user_id: Optional[int]
+    type: Union[pebble.NoticeType, str]
+    key: str
+    first_occurred: datetime.datetime
+    last_occurred: datetime.datetime
+    last_repeated: datetime.datetime
+    occurrences: int
+    last_data: Dict[str, str]
+    repeat_after: Optional[datetime.timedelta]
+    expire_after: Optional[datetime.timedelta]
+
+    def __init__(self, container: Container, id: str, type: str, key: str):
+        self._container = container
+        self.id = id
+        try:
+            self.type = pebble.NoticeType(type)
+        except ValueError:
+            self.type = type
+        self.key = key
+
+        self._notice: Optional[ops.pebble.Notice] = None
+
+    @property
+    def notice(self) -> ops.pebble.Notice:
+        """The non-lazy :class:`ops.pebble.Notice` object (requires a call to Pebble)."""
+        self._ensure_loaded()
+        assert self._notice is not None
+        return self._notice
+
+    def __repr__(self):
+        return f'LazyNotice(id={self.id!r}, type={self.type}, key={self.key!r})'
```

Comment:
Similar to another comment, it feels like it would be worth handling the case where `self.type` is a string here, although this is very much a nit :smile:.

```suggestion
        repr_type = self.type if isinstance(self.type, pebble.NoticeType) else repr(self.type)
        return f'LazyNotice(id={self.id!r}, type={repr_type}, key={self.key!r})'
```

Reply:
Thanks, fixed (and added test).